src/hamiltonian.jl

```diff
@@ -1,8 +1,29 @@
 # TODO: add a type for kinetic energy
 
+# The constructor of `LogDensityValue` will check numerical errors in
+#   `value` and `gradient`.  That is `is_valid` will be performed automatically.
+struct LogDensityValue{Tv<:AbstractFloat, Tg<:AbstractVector{Tv}}
+    value::Tv    # Cached value, e.g. logπ(θ).
+    gradient::Tg # Cached gradient, e.g. ∇logπ(θ).
+end
+
+struct LogDensityFunction{Tf<:Function}
+    f::Tf
+    ∇f::Tf
+end
+
+# The constructor of `PhasePoint` will check numerical errors in
+#   `θ`, `r` and `h`. That is `is_valid` will be performed automatically.
+struct PhasePoint{T<:AbstractVector, Th<:LogDensityValue}
+    θ::T # position variables / parameters
+    r::T # momentum variables
+    h::Th # cached Hamiltonian energy for the current (θ, r)
```

Comment:
This is confusing. Hamiltonian energy doesn't seem to have a type of `LogDensityValue ` right? Should it just be `AbstractFloat`?

Reply:
There's a typo here, `Hamiltonian` energy should be `Potential` energy.